### PR TITLE
feat: updatedRow property should point to the last edited row in table widget

### LINKED ISSUE #15486: [Bug]: Table Widget V2 UpdatedRow Is Always An Empty Object
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

The `Table1.updatedRow` object never gets updated with the edited information when a row is submitted, saved or discarded. The video below for more details

https://user-images.githubusercontent.com/17744578/181265143-90950c30-bccd-467d-8582-82ae2b271dd6.mp4

### Steps To Reproduce

See the video below

https://user-images.githubusercontent.com/17744578/181265143-90950c30-bccd-467d-8582-82ae2b271dd6.mp4

### Public Sample App

_No response_

### Version

Cloud/Appsmith v1.7.9-SNAPSHOT

### PR DESCRIPTION
**Pull request by keyurparalkar**
## Description

This PR updates the logic of how the `updatedRow` derived property is computed. Previously, `updatedRow` was equivalent to `triggeredRow`.  

* ### Intended behaviour:
    * `updatedRow` should be available as soon as a user updates a table cell. Not edits, but updates.
    * `updatedRow` when the mode is set to custom should show data of the most recent row which is updated.
    * The structure of `updatedRow` should be similar to that of `triggeredRow` property. 
    * When there is no change `updatedRow` should be equal to an object with keys having blank string value.

* ### Summary of changes:
    * Introduced a new meta property named: `updatedRowIndex`. This index value helps us to get updated row from the `filteredTableData` or `processTableData`
    * `updatedRowIndex` is updated when the transient table data is updated.
    * It is reset to `-1` whenever:
        * Whenever the value is discarded with discard button. Refer to the `removeRowFromTransientTableData`
        * Whenever the table data is modified. Refer to the `componentDidUpdate`

Fixes #15486

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How Has This Been Tested?

Below is the list of test cases that is executed to verify my changes:
* **Cypress**:
    * Added below cypress tests to `inline_editing_spec.js` to check the following behaviour for single and multi row update mode:
        * check if updatedRowIndex has -1 as the default value (Only for single row update mode)
        * check if updatedRowIndex is 0, when cell at row 0 is updated (Only for single row update mode)
        * check if updatedRowIndex is -1 when changes are discarded
        * check if the updateRowIndex is -1 when widget is reset
        * check if the updatedRowIndex changes to -1 when the table data changes.
* **Unit test**:
    * Added tests in `derived.test.js` with the following scenarios to validate `getUpdatedRow`:
        * with valid updated row index
        * with updated row index as -1
        * with string as a updated row index
        * with undefined as updated row index
        * for removal of non data columns.
* **Manual**
    * Manually tested that, on click of `onSave` button the `updatedRowIndex` is reset to -1.

### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

### QA activity:
- [x] Test plan has been approved by relevant developers
- [x] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/TableV2/Inline_editing_spec.js [hunk 1]
```diff
@@ -711,4 +712,100 @@ describe("Table widget inline editing functionality", () => {
...
+    cy.wait(1000);
+
+    // case 1: check if updatedRowIndex is 0, when cell at row 0 is updated.
+    cy.editTableCell(0, 0);
+    cy.enterTableCellValue(0, 0, "#12").type("{enter}");
+    cy.get(commonlocators.textWidgetContainer).should("contain.text", 0);
+
+    // case 2: check if the updateRowIndex is -1 when widget is reset
+    cy.editTableCell(0, 1);
+    cy.enterTableCellValue(0, 1, "#13").type("{enter}");
```
Comment (sbalaji1192) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/TableV2/Inline_editing_spec.js:798]: we need to check if the updateRowIndex updates to 1 after editing the second row.

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/TableV2/Inline_editing_spec.js [hunk 2]
```diff
@@ -711,4 +712,100 @@ describe("Table widget inline editing functionality", () => {
...
+    cy.contains("Submit").click({ force: true });
+    cy.get(commonlocators.textWidgetContainer).should("contain.text", -1);
+
+    // case 3: check if the updatedRowIndex changes to -1 when the table data changes.
+    cy.wait(1000);
+    cy.editTableCell(0, 2);
+    cy.enterTableCellValue(0, 2, "#14").type("{enter}");
```
Comment (sbalaji1192) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Widgets/TableV2/Inline_editing_spec.js:805]: check here whether updatedRowIndex is 2

**Comment by laveena-en** [comment]
@keyurparalkar 
As discussed, the cursor isn’t seen under editable rows. Please do check.
Also, on discard the updateRow's string is empty. IMO, we could retain the details of the row and revert only the changed value.

Will cover all test cases once the cursor fix is up.

**Comment by keyurparalkar** [comment]
@laveena-en I tested this on the DP. I am not able to reproduce this issue. PFB video 

https://user-images.githubusercontent.com/14138515/203273493-a76586df-29d5-4007-836d-e113a5217f12.mov

**Comment by dilippitchika** [comment]
Ok the reason we started working on this is for single row update mode, once a user discards the change we should always remove the data to not confuse them

**Comment by laveena-en** [comment]
@keyurparalkar Can you try it on a table widget that has Google sheets datasource? Sharing the app for your reference: https://appsmith-git-fix-fix-updatedrow-logic-get-appsmith.vercel.app/app/test12/page1-630481ec84daf727dddb0285

**Comment by keyurparalkar** [comment]
> @keyurparalkar Can you try it on a table widget that has Google sheets datasource? Sharing the app for your reference: https://appsmith-git-fix-fix-updatedrow-logic-get-appsmith.vercel.app/app/test12/page1-630481ec84daf727dddb0285

@laveena-en This is also happening on PROD.

**Comment by laveena-en** [comment]
Raised a separate issue for 

> @keyurparalkar Can you try it on a table widget that has Google sheets datasource? Sharing the app for your reference: https://appsmith-git-fix-fix-updatedrow-logic-get-appsmith.vercel.app/app/test12/page1-630481ec84daf727dddb0285

Raised a separate issue for this https://github.com/appsmithorg/appsmith/issues/18352
cc: @dilippitchika

**Comment by laveena-en** [comment]
Tested and verified for the test-cases mentioned: https://github.com/appsmithorg/TestSmith/issues/2091

**Review by sbalaji1192 — APPROVED** [review]

**Comment by sbalaji1192** [comment]
@keyurparalkar could you update the PR title to include table widget. something like
"feat: updatedRow property should point to the last edited row in table widget"

**Pull request merged**